first commit

ilyador opened the pull request:
I am creating this pull request so I can send you a code review.
Once you fix all the issues, you can mark this pull request as ready to merge into the main branch.
This is how we will work from now on so I can make sure your code is of good quality.

Toriong committed:
first commit

state context providers, pages, and router files added

ilyador commented on src/providers/CartInfoProvider.js:
```diff
@@ -0,0 +1,41 @@
+// all info relating to the cart
+export const CartInfoProvider = (props) => {
+
+    const confirmedCartOrders = JSON.parse(localStorage.getItem("confirmed orders")) === null ? [{
```
This is unreadable. I don't understand the logic of what you were trying to make here.
Please break this functionality down into multiple simple steps.
Once you do that, please move it out of the provider. It is not the provider's job to take care of local storage.
It probably should be done in some to level component like `App`

ilyador commented on src/providers/CartInfoProvider.js:
```diff
@@ -0,0 +1,41 @@
+
+    const cartQuantityTotal = confirmedCartOrders.map((order) => order.confirmedOrderQuantity).reduce((numN, numNMinus1) => numN + numNMinus1);
+
+    const [confirmedOrdersInfo, setConfirmedOrdersInfo] = useState(confirmedCartOrders);
```
There is no point in having 2 states when one is a derivative of another.
Make the calculation once the original state changes.
What is `numNMinus1`? A very unclear name.

ilyador commented on src/providers/CartInfoProvider.js:
```diff
@@ -0,0 +1,41 @@
+        id: null,
+        restaurantName: null,
+        infoOfMainMeatItem: null,
+        confirmedAddOnsToOrder: null,
```
Why is the name `confirmed...`?

ilyador commented on src/providers/CartInfoProvider.js:
```diff
@@ -0,0 +1,41 @@
+    const confirmedCartOrders = JSON.parse(localStorage.getItem("confirmed orders")) === null ? [{
+        id: null,
+        restaurantName: null,
+        infoOfMainMeatItem: null,
```
Try to find shorter ways to name your variables.
This can easily be named `meatItemInfo`

ilyador commented on src/providers/CartInfoProvider.js:
```diff
@@ -0,0 +1,41 @@
+    const [cartTotalPrice, setCartTotalPrice] = useState(cartPriceTotal);
+    const [numberOfCartItems, setNumberOfCartItems] = useState(cartQuantityTotal);
+    // this boolean will be set to true in the navbar component, then, when the SelectedMeatItemViewerToOrderModal component is invoked, as a result of the true boolean, it will compute the price of add-ons that the user selected in their order within a useEffect
+    const [computeConfirmedAddOnsOfCartOrder, setComputeConfirmedAddOnsOfCartOrder] = useState(false);
```
What is the purpose of this state?
Why is the name so long? At least it can be `computeConfirmedCartAddOns`.
This seems to be another state that is derived from the information you hold in another state.

ilyador commented on src/providers/MeatInfoProvider.js:
```diff
@@ -0,0 +1,26 @@
+import React, { useState, createContext, useEffect } from 'react'
+
```
What is the purpose of this provider?
How is it different from the cart?

ilyador commented on src/providers/MeatInfoProvider.js:
```diff
@@ -0,0 +1,26 @@
+export const MeatInfoProvider = (props) => {
+    const [selectedMeatItemInfoFromSearchBar, setSelectedMeatItemInfoFromSeachBar] = useState('');
+    const [selectedAddOns, setSelectedAddOns] = useState([{ name: null, price: 0 }]);
+    const [isRemoveButtonOnDom, setIsRemoveButtonOnDom] = useState(false);
```
We talked about moving this into a display provider.
Why did you create duplicate states?

ilyador commented on src/providers/CartInfoProvider.js:
```diff
@@ -0,0 +1,41 @@
+
+
+    return <CartInfoContext.Provider value={{
+        totalPriceOfCart: [cartTotalPrice, setCartTotalPrice],
```
Why do you not adhere to the same naming pattern?
if you use `cartTotalPrice`, then there is no point naming `totalPriceOfCart`.
You can use `_` for variables that just hold the same thing like this:
`_cartTotalPrice: [cartTotalPrice, setCartTotalPrice]`

ilyador commented on src/providers/MeatInfoProvider.js:
```diff
@@ -0,0 +1,26 @@
+
+
+export const MeatInfoProvider = (props) => {
+    const [selectedMeatItemInfoFromSearchBar, setSelectedMeatItemInfoFromSeachBar] = useState('');
```
This name is way too long (like many others)
Please try to think of how you can describe them in a short and clear way, and which words are simply not necessry.

ilyador commented on src/providers/ModalsConditionalsProvider.js:
```diff
@@ -0,0 +1,25 @@
+
+    const [isOrderMeatItemOrGoToRestaurantMenuModalOpen, setIsOrderMeatItemOrGoToRestaurantMenuModalOpen] = useState(false);
+    const [isCreateNewCartModalOpen, setIsCreateNewCartModalOpen] = useState(false);
+    const [openMeatItemModalFromSearchBar, setOpenMeatItemModalFromSearchBar] = useState(false);
```
If you name all the display states with `is` in the beginning, why not his one?

ilyador commented on src/providers/ModalsConditionalsProvider.js:
```diff
@@ -0,0 +1,25 @@
+    const [isSearchResultsContainerOpen, setIsSearchResultsContainerOpen] = useState(false)
+
+    return <ModalConditionalsContext.Provider value={{
+        isGoToResaurantMenuOrOrderMeatItemModalOpen: [isOrderMeatItemOrGoToRestaurantMenuModalOpen, setIsOrderMeatItemOrGoToRestaurantMenuModalOpen],
```
This variable name is absolutely incomprehensible.

ilyador commented on src/index.js:
```diff
@@ -0,0 +1,24 @@
+
+ReactDOM.render(
+  <React.StrictMode>
+    <ModalsConditionalsProvider>
```
Please separate this into an App component.
You should not putt all this stuff into the render function.

ilyador reviewed (commented):
Please go over all the comments and rename all the variables into shorter and more concise names.

ilyador commented on src/providers/ModalsConditionalsProvider.js:
```diff
@@ -0,0 +1,25 @@
+// the conditionals for opening a modal
+export const ModalsConditionalsProvider = (props) => {
+
+
```
The number of states should correspond directly to the number of modals you have that you can call from different pages. To me it looks like you have more than that.

Toriong committed:
second commit

Toriong committed:
second commit

Toriong committed:
Delete ModalConditionalsAndInfoProvider.js

Toriong committed:
Delete MeatInfoProvider.js

state values are now in 'ModalConditionalsDisplayAndInfoProvider'

Toriong committed:
Delete ModalsConditionalsProvider.js

state values are now in 'ModalConditionalsDisplayAndInfoProvider'

Toriong committed:
second commit

Toriong committed:
second commit

moved all components into the App component.

Toriong committed:
second commit

ilyador commented on src/components/Cart.js:
```diff
@@ -0,0 +1,41 @@
+    // checking and getting the items from local storage and storing them into the state context value of 'confirmedOrdersInfo' onmount
+    // this will check if there is anything in the local storage. If there are values stored in the local storage, then stored them into confirmedOrdersInfo
+    useEffect(() => {
+        if (JSON.parse(localStorage.getItem("confirmed orders") !== null)) {
```
There are 3 different actions here that should be broken down into 3 clear steps:
```js
const _confirmedOrders  = localStorage.getItem("confirmed orders")
const confirmedOrders = JSON.parse(_confirmedOrders)
```
Why are you comparing to `null`? You can just check it is not false
```js
if (confirmedOrders) { ...
```

Toriong replied:
The string "confirmed orders" is the title of the user's order data that is stored in local storage.  `_confirmedOrders` is the keyName that contains `[confirmedOrders, setConfirmedOrders]` from the CartInfoProvider. This state context value will hold the user's cart orders. 

And the reason why I am comparing `null` is that if there's nothing in the local storage, then `JSON.parse(localStorage.getItem("confirmed orders")` will return a `null` value.

ilyador commented on src/components/Cart.js:
```diff
@@ -0,0 +1,41 @@
+    // in order to avoid unnecessary re-renderings by using a state value to get the values below, the useRef will preserve the values through each re-rendering and will update these values when the 'updateCartInfo' is set to true
+    const cartQuantityTotal = useRef(0);
+    const cartPriceTotal = useRef(0);
+
```
You need to display the total value and quantity once that change, so obviously you need a rerender.
I don't understand how `useRef` helps here?
Also, you keep doing what I told you not to do: use 2 different react variables when one is a derivative of the other.

Toriong replied:
So, I can't have any react variables to store the price and quantity total of the cart?  My reasoning for using variables was to display the total price and quantity onto the DOM. I can just put `confirmedOrdersInfo.map((order) => parseFloat(order.totalOrderPrice)).reduce((numA, numB) => numA + numB);` directly into the JSX code in order to display those values. But that seems to be a bit clunky

ilyador replied:
Your cart should be an array, so to show the number of items you should only use `cart.length`

Toriong replied:
But each value--the order--in the array--which will be an object--will have its own quantity.  Each order will be structured as follows: `{
        id: null,
        restaurantName: null,
        infoOfMainMeatItem: null,
        confirmedAddOnsToOrder: null,
        allAddOns: null,
        confirmedOrderQuantity: 0,
        totalMeatPrice: 0,
        totalConfirmedAddOnPrice: 0,
        totalOrderPrice: 0
    }`. So that's why I need to map through the array to get the values under `confirmedOrderQuantity` and find the total sum. That sum will be the total quantity of the cart and will be displayed on the cart icon.

ilyador replied:
>I can just put `confirmedOrdersInfo.map((order) => parseFloat(order.totalOrderPrice)).reduce((numA, numB) => numA + numB);` directly into the JSX code in order to display those values.

No need, you can store it in a normal variable, as those are being regenerated every time a state changes.

ilyador commented on src/providers/ModalConditionalsDisplayAndInfoProvider.js:
```diff
@@ -0,0 +1,38 @@
+import React, { useState, createContext } from 'react'
+
+export const ModalConditionalsDisplayAndInfoContext = createContext()
```
I don't understand what is this file?
Why did you merge the 2 contexts into one?
This is what we talked about that you should separate.

Toriong replied:
The reason why I merged the two files is that all of the values in this file are pertaining to the modal--whether the values are about displaying the modal itself or what's on it. Moreover, on the modal, I am going to have specific buttons displayed (hence, the 'Display' part of the variable name) depending upon the location from where the modal is invoked. With that said, I was going to suggest that maybe I should add another context file for those conditionals about what will be on the modal.

ilyador replied:
> whether the values are about displaying the modal itself or what's on it

These are 2 different things that require different states. If you need to access the same data throughout the app, use another provider.

Also, you already have most of these states in `MeatInfoProvider` which you are duplicating.

> I was going to suggest that maybe I should add another context file for those conditionals about what will be on the modal.

What do they depend on?

Toriong replied:
> These are 2 different things that require different states. If you need to access the same data throughout the app, use another provider.

Ok, I will make another provider then.  And I deleted the MeatInfoProvider and moved some of those values to the ModalConditionalsDisplayAndInfoProvider

> What do they depend on?

If the meat item modal was opened from the cart, then the update button and the remove button will be displayed on the modal. 

And I'm pretty sure we talked about the cart component. We talked about that the cart component and the App component were valid places which I could access the local storage and make the computations for the cart orders.

ilyador replied:
>If the meat item modal was opened from the cart, then the update button and the remove button will be displayed on the modal.

Then those conditionals belong to the cart and you should pass them from the cart to the modal, they are not global.
Try to find a way for each component that brings up the modals to pass its own conditionals, instead of having a Context for them.

Toriong replied:
>Try to find a way for each component that brings up the modals to pass its own conditionals, instead of having a Context for them.

ok, that sounds like a step further down the road--opening the modal from the cart--from where I am right now. Do you want me to focus on building the navbar, the home page, the cart, and the functionality to add items to the cart next?

Toriong replied:
because opening a meat item in a modal assumes at least that I have the menu component coded in.

ilyador reviewed (commented):
You have made 5 commits called `second commit`
Please give your commits meaningful names.
Also, there is no point in creating a commit for just deleting one file.

You still haven't changed a lot of the cumbersome variable names.

ilyador commented:
I am not sure if I should treat this as a finished feature or you are still working on it.
Also, you built the cart component which we did not discuss yet.

Toriong committed:
Renamed ModalConditionals, added new provider

Renamed 'ModalConditionals' to 'IsModalOpenProvider.' 'IsModalOpenProvider' will hold the state values that will determine if the modal is open or not. 
ModalDisplayProvider holds state values that will determine what will be displayed on the modal.

Toriong committed:
App component updated

Toriong committed:
move state context values to Cart.js

Moved the conditionals for the 'del button' and the 'update button' to the Cart component file

Toriong committed:
Delete ModalConditionalsDisplayAndInfoProvider.js

Toriong committed:
Delete ModalDisplayProvider.js

Toriong committed:
Delete ModalDisplayProvider.js
